feat: add `webpackPrefetch`/`webpackPreload`/`webpackFetchPriority` support for URL-based assets

3ru opened the pull request:
# PR Description

This PR refreshes prefetch/preload support across Worker chunks and `new URL()` assets, aligns the behavior with dynamic `import()`, and introduces new runtime helpers.

fixes 
- fixes #19377
- fixes #19373
- fixes #18524

## What kind of change does this PR introduce?

Adds magic comment support to configure resource hints:
- Worker chunks (numeric ordering + fetch priority)
- `new URL()` assets (boolean flags + fetch priority + explicit preload attribute overrides)
- Introduces runtime helpers to inject `<link>` elements at runtime:
- `prefetchAsset` and `preloadAsset`

#### Supported magic comments
- Worker chunks (new Worker / SharedWorker / service worker register):
  - `/* webpackPrefetch: true | number */`
  - `/* webpackPreload: true | number */`
  - `/* webpackFetchPriority: "high" | "low" | "auto" */`
- `new URL()` assets:
  - `/* webpackPrefetch: true */`
  - `/* webpackPreload: true */`
  - `/* webpackFetchPriority: "high" | "low" | "auto" */`
  - `/* webpackPreloadAs: "<request-destination>" */`
  - `/* webpackPreloadType: "<mime/type>" */`
  - `/* webpackPreloadMedia: "<media-query>" */`

Notes:
- Worker accepts number ordering (same as `import()`), URL assets do not.
- Explicit `webpackPreloadAs`/`webpackPreloadType`/`webpackPreloadMedia` override the auto-detected defaults for URL assets.
- The `as` attribute accepts Fetch Standard request destinations. Allowed values include: `audio`, `audioworklet`, `document`, `embed`, `fetch`, `font`, `image`, `manifest`, `object`, `paintworklet`, `report`, `script`, `sharedworker`, `serviceworker`, `style`, `track`, `video`, `worker`, `xslt`.
- Warnings are emitted for invalid values.

### Example Usage

```javascript
// Prefetch an asset for future use
const imageUrl = new URL(
  /* webpackPrefetch: true */
  './assets/hero.jpg',
  import.meta.url
);

// Preload with high priority
const criticalFont = new URL(
  /* webpackPreload: true */
  /* webpackFetchPriority: "high" */
  './fonts/main.woff2',
  import.meta.url
);

// Works with Worker syntax
const worker = new Worker(
  new URL(
    /* webpackPrefetch: true */
    './heavy-computation.worker.js',
    import.meta.url
  )
);

// Override as/type/media explicitly (defaults are auto-detected)
const fontUrl = new URL(
  /* webpackPreload: true */
  /* webpackPreloadAs: "font" */
  /* webpackPreloadType: "font/woff2" */
  /* webpackPreloadMedia: "(max-width: 600px)" */
  './fonts/heading.woff2',
  import.meta.url
);

// Accepts Fetch Request Destinations (e.g., sharedworker)
const helperUrl = new URL(
  /* webpackPreload: true */
  /* webpackPreloadAs: "sharedworker" */
  './scripts/worker-helper.js',
  import.meta.url
);
```

## Did you add tests for your changes?

Yes

## Does this PR introduce a breaking change?

No

## What needs to be documented once your changes are merged?

- Worker: true | number for webpackPrefetch/webpackPreload, fetchpriority supported.
- URL:
  - `webpackPrefetch: true` / `webpackPreload: true`
  - `webpackFetchPriority: "high" | "low" | "auto"`
  - Auto-detected `as` (based on file type), with explicit overrides:
    - `webpackPreloadAs: "<request-destination>"` (Fetch Request Destinations; includes `audio`, `audioworklet`, `document`, `embed`, `fetch`, `font`, `image`, `manifest`, `object`, `paintworklet`, `report`, `script`, `sharedworker`, `serviceworker`, `style`, `track`, `video`, `worker`, `xslt`)
    - `webpackPreloadType: "<mime/type>"`
    - `webpackPreloadMedia: "<media-query>"`
  - Invalid values emit warnings.


---
rel

- #17121

3ru committed:
feat: add webpackPrefetch/webpackPreload/webpackFetchPriority support for Worker syntax

alexander-akait commented on test/configCases/worker/worker-prefetch-preload/index.js:
```diff
@@ -0,0 +1,219 @@
+	const worker = new Worker(new URL("./prefetch-worker.js", import.meta.url), {
+		type: "module",
+		/* webpackPrefetch: true */
+	});
```
Why not:
```js
const worker = new Worker(/* webpackPrefetch: true */ new URL("./prefetch-worker.js", import.meta.url), {
		type: "module"
});
```

We usually use comments as a prefix.

3ru committed:
test: update Worker magic comment tests to follow webpack conventions

alexander-akait commented on test/configCases/worker/worker-prefetch-preload/index.js:
```diff
@@ -0,0 +1,293 @@
+	});
+	expect(result).toBe("prefetch-worker: test");
+	await worker.terminate();
+});
```
Same for preload

3ru replied:
@alexander-akait You're right. I understand the issue now.

## Implementation Plan

### Step 1: Implement `new URL()` prefetch/preload support first

As you suggested, this is the more general solution that benefits all URL-based modules:

```javascript
new URL(
  /* webpackPrefetch: true */
  /* webpackPreload: 3 */
  "./heavy-asset.wasm",
  import.meta.url
)
```


### Step 2: Add Worker integration on top of URL support

Once `new URL()` supports prefetch/preload, Worker support becomes straightforward:

```javascript
new Worker(
  /* webpackPrefetch: true */
  new URL("./worker.js", import.meta.url)
)
```

- Emit warnings when both are specified
- Maintain backward compatibility

### Step 3: Handle edge cases and conflicts

```javascript
// Conflict resolution
new Worker(
  /* webpackPrefetch: 1 */  // This wins
  new URL(
    /* webpackPrefetch: 2 */ // This is ignored with warning
    "./worker.js",
    import.meta.url
  )
)
```

### Addressing the timing concern (before/after chunk loading)

For the timing issue that blocked #18524, I propose adding an explicit control:

```javascript
new URL(
  /* webpackPrefetch: true */
  /* webpackPrefetchTiming: "immediate" | "afterLoad" | "onIdle" */
  "./asset.wasm",
  import.meta.url
)
```

- **immediate**: Start prefetching as soon as the parent chunk loads (before)
- **afterLoad**: Wait until the URL is actually used once (after)
- **onIdle**: Use requestIdleCallback (default behavior)

Would this approach address the timing concerns?

alexander-akait replied:
I think we don't need syntax for `new Worker` at all, only for `new URL`, because it has not sense, 

https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Attributes/rel/preload comments for reload (i.e. type, as, crossorigin and etc like `webpackPreloadType`, `webpackPreloadAs` and etc)
https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Attributes/rel/prefetch comments for prefetch

Shorty - we should emulate HTML `link[prefetch]` and `link[preload]` behavior here

About when we should load it - it is good question, I think loading them after loading chunk by default is good place todo it, let's wait an answer from @hai-x @xiaoboost 

Anyway you can start to implement it after chunk loading, it is not hard to change it or implement behavior `webpackPrefetchTiming`, need a discussion here.

Also feel free to reuse my work from linked PR

3ru replied:
After trial and error, I've successfully implemented the feature for `new URL()` syntax as you suggested:

```javascript
new URL(
	/* webpackPrefetch: true */
	/* webpackPreload: true */
	/* webpackFetchPriority: "high" */
	"./asset.png",
	import.meta.url
)
```

This approach is indeed more general and useful - it works for any `new URL()` modules including assets, and naturally extends to Worker usage:

```ts
new Worker(
	new URL(/* webpackPrefetch: true */ "./worker.js", import.meta.url)
)
```

I realize now our initial discussion had some crossed wires about the implementation approach.

3ru committed:
refactor: move prefetch/preload support from Worker to new URL() syntax

3ru committed:
feat: implement URL-based prefetch/preload support with runtime modules

3ru committed:
feat: implement webpackPrefetch/webpackPreload/webpackFetchPriority support for new URL() syntax

3ru committed:
fix: resolve linting errors in test files

xiaoxiaojx commented on lib/dependencies/URLDependency.js:
```diff
@@ -132,7 +140,108 @@ URLDependency.Template = class URLDependencyTemplate extends (
+				// Reference: MDN rel=preload documentation (https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Attributes/rel/preload)
+				// Valid 'as' values: fetch, font, image, script, style, track
+				// Note: audio/video are in spec but not supported by major browsers as of 2025
+				if (/\.(png|jpe?g|gif|svg|webp|avif|bmp|ico|tiff?)$/i.test(request)) {
```
Consider extracting this into a separate function.

3ru replied:
fixed in bce3205c79214b02a855179798bac7e5de8d393e

xiaoxiaojx commented on lib/runtime/AssetPrefetchPreloadRuntimeModule.js:
```diff
@@ -0,0 +1,65 @@
+				// Add nonce if needed
+				compilation.outputOptions.crossOriginLoading
+					? Template.asString([
+							"if(__webpack_require__.nc) {",
```
Use RuntimeGlobals.scriptNonce instead.

3ru replied:
fixed in 8daacd42f3494bd06ea1a801f31ec070c1a730c1

alexander-akait commented on lib/url/URLParserPlugin.js:
```diff
@@ -184,6 +184,86 @@ class URLParserPlugin {
+								/** @type {DependencyLocation} */ (expr.loc)
+							)
+						);
+					}
```
We'll also need `webpackPreloadType` - ideally we'll have the same attributes as `link` - https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/link

3ru replied:
added in e7f4a0d32ac69ea77fc7fbe8b5a51723bd70d8a9

alexander-akait commented on lib/url/URLParserPlugin.js:
```diff
@@ -184,6 +184,86 @@ class URLParserPlugin {
+								/** @type {DependencyLocation} */ (expr.loc)
+							)
+						);
+					}
```
Avoid it, we don't use this in other places, we need to discuss about it and if we want to validate this we should add this to all places (i.e. `import()`, `Worker`, `require()`, etc), because there are other comments for other comments that override each other

3ru replied:
> because there are other comments for other comments that override each other

I don't understand what this means.

When both `prefetch` and `preload` are specified like in this case, what kind of link do you think would be ideal to generate?

alexander-akait commented on test/configCases/asset-modules/url-prefetch-preload-fetchpriority/index.js:
```diff
@@ -0,0 +1,170 @@
+		};
+		return element;
+	}
+};
```
We have `moduleScope` function to inject something in global scope, or feel free to improve our DOM mock for tests, you can use another PR for this

3ru replied:
fixed in c632fcc9eef369f2b1f7deddefc464427a748e4b

alexander-akait commented on test/dependencies/URLDependency.unittest.js:
```diff
@@ -0,0 +1,212 @@
+			});
+		});
+	});
+});
```
Avoid unit tests, they are useless in term of webpack testing and DOM accessing , integration tests take more time but they are more valuable and allow you to catch regressions

3ru replied:
removed in c632fcc9eef369f2b1f7deddefc464427a748e4b

xiaoxiaojx commented on lib/dependencies/URLDependency.js:
```diff
@@ -132,7 +140,108 @@ URLDependency.Template = class URLDependencyTemplate extends (
+					dep.outerRange[1] - 1,
+					`/* asset import */ (function() {
+						var url = new ${RuntimeGlobals.relativeUrl}(${moduleExpr});
+						${hintCode.join("\n")}
```
We’re loading it too late here, which defeats the purpose of prefetch.

3ru replied:
@xiaoxiaojx Thank you for the excellent catch!

I've implemented a fix [`61bed79`](https://github.com/webpack/webpack/pull/19695/commits/61bed79755bd1a12f10a7b44b259d293ef1d3a09) that moves asset prefetch/preload execution to startup time:

### Changes made:
  1. Created `AssetPrefetchStartupPlugin`: Collects all assets with prefetch/preload directives at compile time
  2. Created `AssetPrefetchStartupRuntimeModule`: Generates runtime code that executes at startup using `onChunksLoaded` (same pattern as chunk prefetch)
  3. Modified `URLDependency`: Added support for startup execution mode while maintaining backward compatibility

I believe this properly addresses your concern about timing.

3ru replied:
Could you check this change?
825fec525f94fc861b560bd477f02ec19b542599


1. Compile-time collection: The `AssetPrefetchStartupPlugin` collects all URLDependencies with prefetch/preload during the finishModules hook and marks them.
2. Startup-time execution: Instead of using inline IIFEs, the prefetch/preload hints are now executed at module startup through `AssetPrefetchStartupRuntimeModule` with
`RuntimeModule.STAGE_TRIGGER`.

3ru committed:
refactor: remove prefetch/preload conflict warning and use RuntimeGlobals.scriptNonce

3ru committed:
refactor: extract asset type determination logic into separate function

3ru committed:
test: clean up tests and remove outdated functionality

3ru committed:
feat: add webpackPreloadType support for new URL() syntax

3ru committed:
fix: move asset prefetch/preload execution from inline to startup time

3ru committed:
fix: resolve cache invalidation issues in asset prefetch implementation

3ru committed:
refactor: extract asset type determination logic into util module

3ru committed:
chore: comments

3ru committed:
refactor: simplify asset prefetch/preload implementation

- Consolidate multiple runtime modules into unified AssetResourcePrefetchPlugin
- Remove complex startup prefetch mechanism in favor of simpler inline approach

3ru commented:
rebased

xiaoxiaojx commented on lib/asset/AssetResourcePrefetchPlugin.js:
```diff
@@ -0,0 +1,59 @@
+					set.add(RuntimeGlobals.require);
+					set.add(RuntimeGlobals.baseURI);
+					if (set.has(RuntimeGlobals.relativeUrl)) {
+						set.add(RuntimeGlobals.relativeUrl);
```
typo ?

3ru committed:
fix: only enable URL asset prefetch/preload hints for web targets

3ru commented on lib/url/URLParserPlugin.js:
```diff
@@ -184,6 +184,49 @@ class URLParserPlugin {
+
+					// Accept only boolean true for webpackPreload
+					if (importOptions.webpackPreload === true) {
+						dep.preload = true;
```
right! fixed e7c0382087e9ace904e9a2006f2bf9f09749a754

alexander-akait commented on lib/url/URLParserPlugin.js:
```diff
@@ -184,6 +185,51 @@ class URLParserPlugin {
+								/** @type {DependencyLocation} */ (expr.loc)
+							)
+						);
+					}
```
We need to implement `webpackPreloadAs`, `webpackPreloadType` and  `webpackPreloadMedia` - https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Attributes/rel/preload

3ru replied:
Just to clarify your suggestion, are you proposing that we should:

1. Keep the automatic detection (like in assetType.js) as the default behavior
2. But also allow developers to explicitly override these attributes via magic comments like `webpackPreloadAs`, `webpackPreloadType`, and `webpackPreloadMedia`

So essentially, automatic detection as a default, with manual overrides when needed?

alexander-akait commented on lib/util/assetType.js:
```diff
@@ -0,0 +1,34 @@
+		return "fetch";
+	}
+	return "fetch";
+};
```
Make it as a private method of [AssetResourcePrefetchRuntimeModule.js](https://github.com/webpack/webpack/pull/19695/files#diff-662ac35ab1c4e1186e28037ff997e326ae043137871ff5352cc9441a1b73777d), we cache runtime generation, so we don't need to store it in dependecies

3ru replied:
fixed by 719e2b13ca77aa9d5c44ac19d9f46a27c702b193

alexander-akait reviewed (commented):
@3ru let's do some changes

3ru committed:
refactor: replace `AssetResourcePrefetchRuntimeModule` with `ResourcePrefetchRuntimeModule`

3ru committed:
refactor: move asset type determination logic from util module to URLDependency class

3ru commented on lib/asset/AssetResourcePrefetchRuntimeModule.js:
```diff
@@ -0,0 +1,82 @@
+	}
+}
+
+module.exports = AssetResourcePrefetchRuntimeModule;
```
fixed by 0e3dfa1a38ae64d6c40d59ee829fe75473f32cfd

3ru committed:
feat(runtime): improve asset prefetch/preload for universal target

3ru commented on lib/dependencies/URLPlugin.js:
```diff
@@ -50,7 +50,14 @@ class URLPlugin {
+					new URLParserPlugin({
+						...parserOptions,
+						enableUrlAssetHints
+					}).apply(parser);
```
Does this change align with your ideals? 37632364c5e3b02d05b051e9c9ba269d8991add2

alexander-akait replied:
Just add a runtime check, like we have for CSS, because we need to generate code when we are in web env

3ru committed:
feat: add preload options for enhanced asset loading

- Introduced `preloadAs`, `preloadType`, and `preloadMedia` properties to `URLDependency` for better control over asset preloading.

3ru committed:
tests: add test cases for invalid `webpackPreloadAs`, `webpackPreloadType`, and `webpackPreloadMedia` values.

3ru committed:
refactor

3ru committed:
refactor: simplify `webpackPreloadAs` validation and update warnings

3ru commented on lib/url/URLParserPlugin.js:
```diff
@@ -224,6 +224,75 @@ class URLParserPlugin {
+					// webpackPreloadAs: allow override of the "as" attribute for preload
+					// Fetch Standard: Request destinations (enumerates destinations used by `as`) https://fetch.spec.whatwg.org/#concept-request-destination
+					if (importOptions.webpackPreloadAs !== undefined) {
+						const allowedAs = [
```
Simplified a8d39c0cdfe243ee8111d8114895eaddfdc197d4

PupilTong commented:
How about the `webpackChunkName` ?

alexander-akait commented:
@PupilTong We don't have `webpackChunkName` here, because we are working with modules, not chunks, technically we can introduce `webpackModuleName`, but it is out of scope this PR, here only prefetch/preload

evenstensberg commented:
Let's try to get this reviewd, patched and merged soon.

alexander-akait commented:
Let’s close in favor of https://github.com/webpack/webpack/pull/21012, now we solve it for all module types and allow to setup for many cases, anyway thanks for the pr
